types/networks/src/lib.rs

```diff
@@ -5,204 +5,440 @@
+
+#[derive(Clone)]
+struct DrandPoint<'a> {
+    pub height: ChainEpoch,
+    pub config: &'a DrandConfig<'a>,
+}
+
+/// Defines all network configuration parameters.
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+#[serde(default)]
+pub struct ChainConfig {
+    pub name: String,
+    pub bootstrap_peers: Vec<String>,
+    pub block_delay_secs: u64,
+    pub version_schedule: Vec<UpgradeInfo>,
+    pub height_infos: Vec<HeightInfo>,
 }
 
-/// Constructs a drand beacon schedule based on the build config.
-pub async fn beacon_schedule_default(
-    genesis_ts: u64,
-) -> Result<BeaconSchedule<DrandBeacon>, Box<dyn Error>> {
-    let mut points = BeaconSchedule(Vec::with_capacity(DRAND_SCHEDULE.len()));
-    for dc in DRAND_SCHEDULE.iter() {
-        points.0.push(BeaconPoint {
-            height: dc.height,
-            beacon: Arc::new(DrandBeacon::new(genesis_ts, BLOCK_DELAY_SECS, dc.config).await?),
-        });
+impl ChainConfig {
+    pub fn calibnet() -> Self {
+        use calibnet::*;
+        Self {
+            name: "calibnet".to_string(),
+            bootstrap_peers: DEFAULT_BOOTSTRAP.iter().map(|x| x.to_string()).collect(),
+            block_delay_secs: EPOCH_DURATION_SECONDS as u64,
+            version_schedule: UPGRADE_INFOS.to_vec(),
+            height_infos: HEIGHT_INFOS.to_vec(),
+        }
+    }
+
+    pub fn conformance() -> Self {
+        Self {
+            height_infos: CONFORMANCE_HEIGHT_INFOS.to_vec(),
+            ..Self::default()
+        }
+    }
+
+    pub fn network_version(&self, epoch: ChainEpoch) -> NetworkVersion {
+        let height = self
+            .height_infos
+            .iter()
+            .rev()
+            .find(|info| epoch > info.epoch)
+            .map(|info| info.height)
+            .unwrap_or(Height::Breeze);
+
+        self.version_schedule
+            .iter()
+            .find(|info| height == info.height)
+            .map(|info| info.version)
+            .expect("A network version should exist even if not specified in the config (a default exists).")
+    }
+
+    pub async fn get_beacon_schedule(
+        &self,
+        genesis_ts: u64,
+    ) -> Result<BeaconSchedule<DrandBeacon>, Box<dyn Error>> {
+        let ds_iter = if self.name == "calibnet" {
+            calibnet::DRAND_SCHEDULE.iter()
+        } else {
+            mainnet::DRAND_SCHEDULE.iter()
+        };
+        let mut points = BeaconSchedule(Vec::with_capacity(ds_iter.len()));
+        for dc in ds_iter {
+            points.0.push(BeaconPoint {
+                height: dc.height,
+                beacon: Arc::new(
+                    DrandBeacon::new(genesis_ts, self.block_delay_secs, dc.config).await?,
+                ),
+            });
+        }
+        Ok(points)
+    }
+
+    pub fn epoch(&self, height: Height) -> ChainEpoch {
+        self.height_infos
+            .iter()
+            .find(|info| height == info.height)
+            .map(|info| info.epoch)
+            .unwrap()
```

Comment:
perhaps an `expect` with a nice description would be more gentle in a terrible case where it may fail (or perhaps it can't?)

Reply:
Done. It would be nice to prove that it cannot happen.

Reply:
@elmattic how about adding a shiny, low-priority issue with a nice description to have this validation so someone could take a look in a separate PR?